### PR TITLE
Fix custom expression simplification

### PR DESCRIPTION
**Pull request by clinssen**
- Fix custom expression simplification not being applied consistently
- Introduce a global configuration data structure to centralize global options

### CODE REVIEW

File: odetoolbox/__init__.py [hunk 1]
```diff
@@ -110,12 +95,13 @@ def _from_json_to_shapes(indict, options_dict, parameters=None) -> List[Shape]:
     all_parameter_symbols = set()
     all_variable_symbols_ = set()
     for shape_json in indict["dynamics"]:
-        shape = Shape.from_json(shape_json, time_symbol=options_dict["input_time_symbol"], differential_order_symbol=options_dict["differential_order_symbol"], parameters=parameters)
+        shape = Shape.from_json(shape_json, time_symbol=Config().input_time_symbol, parameters=parameters)
```
Comment (pnbabu) [inline odetoolbox/__init__.py:98]: The `time_symbol` argument isn't used in the `from_json` function. Can we remove it?

File: odetoolbox/__init__.py [hunk 2]
```diff
@@ -110,12 +95,13 @@ def _from_json_to_shapes(indict, options_dict, parameters=None) -> List[Shape]:
...
         all_variable_symbols.extend(shape.get_state_variables())
-        all_variable_symbols_.update(shape.get_state_variables(derivative_symbol="__d"))
+        all_variable_symbols_.update(shape.get_state_variables(derivative_symbol=Config().differential_order_symbol))
```
Comment (pnbabu) [inline odetoolbox/__init__.py:100]: Do we need to pass the `Config` options here as it is a global class that can be accessed directly in the function as well? This applies to all the places where `Config` is passed as a function argument.

File: odetoolbox/shapes.py [hunk 3]
```diff
@@ -596,24 +583,25 @@ def from_ode(cls, symbol: str, definition: str, initial_values: dict, all_variab
...
-        derivative_factors, inhom_term, nonlin_term = Shape.split_lin_inhom_nonlin(definition, all_variable_symbols, parameters=parameters)
+            all_variable_symbols.extend(local_symbols_sympy)
+        all_variable_symbols = [str(sym_name).replace("'", Config().differential_order_symbol) for sym_name in all_variable_symbols]
+        all_variable_symbols_sympy = [sympy.Symbol(sym_name) for sym_name in all_variable_symbols]
+        derivative_factors, inhom_term, nonlin_term = Shape.split_lin_inhom_nonlin(definition, all_variable_symbols_sympy, parameters=parameters)
         local_symbols_idx = [all_variable_symbols.index(sym) for sym in local_symbols]
         local_derivative_factors = [derivative_factors[i] for i in local_symbols_idx]
-        nonlocal_derivative_terms = [derivative_factors[i] * all_variable_symbols[i] for i in range(len(all_variable_symbols)) if i not in local_symbols_idx]
+        nonlocal_derivative_terms = [derivative_factors[i] * sympy.Symbol(all_variable_symbols[i]) for i in range(len(all_variable_symbols)) if i not in local_symbols_idx]
```
Comment (pnbabu) [inline odetoolbox/shapes.py:600]: Can you use `all_variable_symbols_sympy` here instead?

File: odetoolbox/__init__.py (hunk 2 above)
Comment (clinssen) [reply odetoolbox/__init__.py:100]: Thanks for noticing this, I cleaned it up in another place, but in the remainder of cases ``Shape.get_state_variables()`` is called with different values of this parameter.

File: doc/index.rst [hunk 4]
```diff
@@ -348,7 +345,11 @@ The following global options are defined. Note that all are typically formatted
    * - ``differential_order_symbol``
      - :python:`"__d"`
      - string
-     - String appended n times to output variable names to indicate differential order n. TODO: only the default value works for now.
+     - String appended n times to output variable names to indicate differential order n.
+   * - ``simplify_expression``
+     - :python:`"sympy.logcombine(sympy.powsimp(sympy.expand(expr)))"`
```
[the remark this answers is not in the document]
Comment (clinssen) [reply doc/index.rst:350]: This was part of performance optimization tests for NESTML, but if we want to override the default we should probably do that from the NESTML side. I have changed it back for now.

**Pull request merged**